### PR TITLE
Payment

### PR DESCRIPTION
**Pull request by nalijagic1**
This pull request contains implementation of payment feature. Payment consists of 2 steps: address informatio and card information

### CODE REVIEW

File: backend/src/main/java/com/praksa/auction/config/security/services/PersonDetails.java [hunk 1]
```diff
@@ -128,12 +127,12 @@ public void setAddress(Address address) {
...
 
-    public Card getCard() {
-        return card;
+    public String getCustomerId() {
+        return customerId;
     }
 
-    public void setCard(Card card) {
-        this.card = card;
+    public void setCustomerId(String customerId) {
```
Comment (neirapulo-abh) [inline backend/src/main/java/com/praksa/auction/config/security/services/PersonDetails.java]: Use lombok

File: backend/src/main/java/com/praksa/auction/model/Person.java [hunk 2]
```diff
@@ -123,12 +121,12 @@ public void setAddress(Address address) {
         this.address = address;
     }
 
-    public Card getCard() {
-        return card;
-    }
 
-    public void setCard(Card card) {
-        this.card = card;
+    public String getCustomerId() {
```
Comment (neirapulo-abh) [inline backend/src/main/java/com/praksa/auction/model/Person.java]: Use lombok

File: backend/src/main/java/com/praksa/auction/model/Product.java [hunk 3]
```diff
@@ -140,4 +141,11 @@ public void setAddress(Address address) {
         this.address = address;
     }
 
+    public Boolean getPayed() {
```
Comment (neirapulo-abh) [inline backend/src/main/java/com/praksa/auction/model/Product.java]: Use lombok

File: backend/src/main/java/com/praksa/auction/model/Country.java [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+    public Country(long id, String name) {
+        this.id = id;
+        this.name = name;
+    }
+
+    public Country() {
+    }
+
+    public long getId() {
```
Comment (neirapulo-abh) [inline backend/src/main/java/com/praksa/auction/model/Country.java]: Use lombok

**Review by neirapulo-abh — CHANGES_REQUESTED** [review]

File: backend/src/main/java/com/praksa/auction/controller/CountryController.java [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+public class CountryController {
+    private final CountryService countryService;
+
+    @Autowired
+    public CountryController(CountryService countryService) {
+        this.countryService = countryService;
+    }
+
+    @GetMapping()
+    public ResponseEntity<List<Country>> getAllConutries() {
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/controller/CountryController.java]: getAllConutries -> getAllCountries

File: backend/src/main/java/com/praksa/auction/controller/PaymentController.java [hunk 6]
```diff
@@ -0,0 +1,25 @@
...
+@RestController
+public class PaymentController {
+    @Autowired
+    private final PaymentService paymentService;
+
+    public PaymentController(PaymentService paymentService) {
+        this.paymentService = paymentService;
+    }
+
+    @PostMapping("/payment/paymentIntent")
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/controller/PaymentController.java]: Add '/payment' above Controller

File: backend/src/main/java/com/praksa/auction/controller/PaymentController.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    public ResponseEntity<?> createPaymentIntente(@RequestBody PaymentInfoDto paymentInfoDto) throws StripeException {
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/controller/PaymentController.java]: Please remove ? - use defined typer for arguments and method return types.

File: backend/src/main/java/com/praksa/auction/controller/ProductController.java [hunk 8]
```diff
@@ -59,4 +59,10 @@ public ResponseEntity<Product> getProductById(@PathVariable int id) {
         Optional<Product> product = productService.getSelectedProduct(id);
         return ResponseEntity.ok(product.get());
     }
+
+    @PutMapping("/updatePayedStatus")
+    public ResponseEntity<?> updatePayedStatus(@RequestBody boolean payed, @RequestParam long product) {
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/controller/ProductController.java]: Remove ?

File: backend/src/main/java/com/praksa/auction/model/Person.java [hunk 9]
```diff
@@ -21,114 +30,5 @@ public class Person {
     @ManyToOne
     @JoinColumn(name = "addressId")
     private Address address;
-    @OneToOne(fetch = FetchType.LAZY)
-    @JoinColumn(name = "cardId")
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/model/Person.java:25]: Drop table Card and this column cardId from Person table.

File: backend/src/main/java/com/praksa/auction/repository/PersonRepository.java [hunk 10]
```diff
@@ -2,10 +2,17 @@
...
 
 public interface PersonRepository extends JpaRepository<Person, Long> {
     Optional<Person> findByEmail(String email);
 
     Boolean existsByEmail(String email);
+
+    @Query(value = "UPDATE Person SET customer_id = :customerId WHERE id = :userId", nativeQuery = true)
+    void updateCustomerInfo(String customerId, long userId);
+
+    @Query(value = "UPDATE Person SET address_id = :addressId WHERE id = :userId", nativeQuery = true)
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/repository/PersonRepository.java:23]: Add Transactional  and Modifying annotation and test it again.

File: backend/src/main/java/com/praksa/auction/service/AddressService.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    private final PersonService personService;
+
+    public AddressService(AddressRepository addressRepository, CountryService countryService, PersonService personService) {
+        this.addressRepository = addressRepository;
+        this.countryService = countryService;
+        this.personService = personService;
+    }
+
+    private Address createAddressFromRequest(AddressInfoDto addressInfoDto) {
+        Address address = new Address();
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/service/AddressService.java]: Create contructor in Address model with four arguments added below.

File: backend/src/main/java/com/praksa/auction/service/AddressService.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+        address.setStreet(addressInfoDto.getAddress());
+        address.setCity(addressInfoDto.getCity());
+        address.setCountry(countryService.findById(addressInfoDto.getCountryId()));
+        address.setZipCode(addressInfoDto.getZipCode());
+        return address;
+    }
+
+    public Address addAddressIfNotExist(AddressInfoDto address, long personId) {
+        Address location = createAddressFromRequest(address);
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/service/AddressService.java]: Maybe add address instaed of location

File: backend/src/main/java/com/praksa/auction/service/PaymentService.java [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+@Service
+public class PaymentService {
+    @Autowired
+    PersonService personService;
+    @Value("${stripeSecretKey}")
+    private String apiKey;
+
+    public PaymentResponseDto createPaymentIntent(PaymentInfoDto paymentInfoDto) throws StripeException {
+        Person buyer = personService.getPersonById(paymentInfoDto.getCustomerId());
+        String custumerId = "";
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/service/PaymentService.java]: String custumerId = buyer.getCustomerId() ? personService.createCustomerId(buyer): buyer.getCustomerId();

File: backend/src/main/java/com/praksa/auction/service/PaymentService.java [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+                PaymentIntentCreateParams.builder()
+                        .setAmount((long) (paymentInfoDto.getAmount() * 100))
+                        .setCurrency("USD")
+                        .setCustomer(custumerId)
+                        .addPaymentMethodType("card")
+                        .putMetadata("product_id", String.valueOf(paymentInfoDto.getProductId()))
+                        .build();
+        PaymentIntent paymentIntent = PaymentIntent.create(params);
+        PaymentResponseDto paymentResponse = new PaymentResponseDto(paymentIntent.getClientSecret());
+        return paymentResponse;
```
Comment (eminaomanovic) [inline backend/src/main/java/com/praksa/auction/service/PaymentService.java]: return new PaymentResponseDto(paymentIntent.getClientSecret());

File: backend/src/test/java/com/praksa/auction/BiddingTest.java [hunk 15]
```diff
@@ -0,0 +1,27 @@
...
+public class BiddingTest {
+    @Autowired
+    BidController bidController;
+    @Autowired
+    BidRepository bidRepository;
+    @Test
+    void bidding() {
+        Bid highestBidOld =  bidRepository.findFirstByProductIdOrderByBidDesc(5);
+        bidController.placeBid(new BiddingInfoDto(5,2,35.50));
+        Bid highesrBidNew = bidRepository.findFirstByProductIdOrderByBidDesc(5);
```
Comment (eminaomanovic) [inline backend/src/test/java/com/praksa/auction/BiddingTest.java]: highesrBidNew -> highestBidNew

File: frontend/src/components/auctionItem/auctionItem.css [hunk 16]
```diff
@@ -0,0 +1,66 @@
...
+  font-family: "Lato";
+  font-style: normal;
+  font-weight: 400;
+  font-size: 18px;
+  line-height: 27px;
+  color: #252525;
+}
+
+.auctionPrice {
+  float: right;
```
Comment (eminaomanovic) [inline frontend/src/components/auctionItem/auctionItem.css]: Inline css ?

File: frontend/src/components/auctionItem/auctionItem.css [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+  font-family: "Lato";
+  font-style: normal;
+  font-weight: 400;
+  font-size: 16px;
+  line-height: 24px;
+  color: #8367d8;
+}
+
+.auctionItem h2 {
+  font-family: "Lato";
```
Comment (eminaomanovic) [inline frontend/src/components/auctionItem/auctionItem.css]: Try to import font-family as separated class for elements.

File: frontend/src/components/notification/notification.css [hunk 18]
```diff
@@ -15,6 +15,10 @@
   color: #417505;
 }
 
+.notificationinfo {
+  color: #8367d8;
+  background-color: #f1effa;
+}
 .notificationwarning {
```
Comment (eminaomanovic) [inline frontend/src/components/notification/notification.css]: notification-info, notification-warning...

File: frontend/src/components/paymentInfo/paymentInfo.css [hunk 19]
```diff
@@ -0,0 +1,25 @@
+.paymentInformation {
```
Comment (eminaomanovic) [inline frontend/src/components/paymentInfo/paymentInfo.css:1]: paymentInformation -> payment-information

File: frontend/src/components/paymentOverview/paymentOverview.js [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+import { useNavigate } from "react-router-dom";
+import Checkbox from "../checkbox/checkbox";
+import AuctionItem from "../auctionItem/auctionItem";
+function PaymentOverview(props) {
+  const navigate = useNavigate();
+  return (
+    <div className="paymentOverview">
+      <h5>REVIEW AND CONFIRM PAYMENT</h5>
+      <div className="reviewContent">
+        <Dropdown title="Shipping Address" content={[<p>Uslo</p>]} />
```
Comment (eminaomanovic) [inline frontend/src/components/paymentOverview/paymentOverview.js]: Remove Uslo info and additional view - which is not needed.

File: frontend/src/components/paymentOverview/paymentOverview.js [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+          <Dropdown
+            title="Shipping Address"
+            content={[
+              <p>{props.location.address}</p>,
+              <p>{props.location.city}</p>,
+              <p>{props.location.country}</p>,
+              <p>{props.location.zipCode}</p>,
+              <p className="change" onClick={() => props.goToStep(1)}>
+                Change
+              </p>,
```
Comment (eminaomanovic) [inline frontend/src/components/paymentOverview/paymentOverview.js:24]: Please check if comma is needed here.

File: frontend/src/components/shippingAddress/shippingAddress.js [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+  return (
+    <div className="shippingAddress">
+      <h5>SHIPPING ADDRESS</h5>
+      <form>
+        <Field
+          placeHolder="Enter your address"
+          fieldClass="loginAndRegisterField"
+          label="Address*"
+          type="text"
+          id="addrss"
```
Comment (eminaomanovic) [inline frontend/src/components/shippingAddress/shippingAddress.js]: addrss -> address

File: frontend/src/components/shippingAddress/shippingAddress.js [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+          onChange={(e) => {
+            setError({
```
Comment (eminaomanovic) [inline frontend/src/components/shippingAddress/shippingAddress.js]: Add method for setting error above - updateErrorMessage(field, message).

File: frontend/src/pages/payment/payment.js [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+import { loadStripe } from "@stripe/stripe-js";
+import paymentService from "../../services/payment.service";
+import personService from "../../services/person.service";
+
+function Payment() {
+  const { state } = useLocation();
+  const { price, product } = state;
+  const [currentStep, setCurrentStep] = useState(0);
+  const [enteredAddress, setEnteredAddress] = useState();
+  const stripePromise = loadStripe(
```
Comment (eminaomanovic) [inline frontend/src/pages/payment/payment.js]: Please add this to constants.js file.

File: frontend/src/pages/payment/payment.js [hunk 25]
```diff
@@ -0,0 +1,71 @@
...
+    "pk_test_51LVB8xBj1vinbdx6NyD5IYpzSgWonLR41HNAETguKGEXelw3DcPqP0l3JQ69NAsJOjrfPue0tK2rjJ423fr5PgJS00fSun4phx"
+  );
+  const [clientSecret, setClientSecret] = useState();
+  const appearance = {
+    theme: "stripe",
+  };
+  const options = {
```
Comment (eminaomanovic) [inline frontend/src/pages/payment/payment.js:20]: const options = {client, {
    theme: "stripe",
  }};

File: frontend/src/utils/notificationConstants.js [hunk 26]
```diff
@@ -8,4 +8,7 @@ export const NOTIFICATION_MESSAGES = {
   BID_HIGHER_MESSAGE:
     "There are higher bids than yours. You could give a second try!",
   NO_BID: "Please enter a number in order to place your bid",
+  PAYMENT_FINISHED: "You have successfully completed payment for this product",
+  AUCTION_ENDED: "This auction has ended",
+  OUTBID: "Congratulations! You outbid the competition.",
```
Comment (eminaomanovic) [inline frontend/src/utils/notificationConstants.js:13]: Please add all messages here.

File: frontend/src/utils/validation.js [hunk 27]
```diff
@@ -78,6 +113,39 @@ class Validation {
...
+        case "city":
+          errorMessages.city = this.validateCity(location.city);
+          break;
+        case "zipCode":
+          errorMessages.zipCode = this.validateZipCode(location.zipCode);
+          break;
+        case "countryId":
+          errorMessages.country = this.validateCountry(location.countryId);
+          break;
+        default:
```
Comment (eminaomanovic) [inline frontend/src/utils/validation.js:133]: Please add this to default case:
errorMessages.country = this.validateCountry(location.countryId);
          break;

File: backend/src/main/java/com/praksa/auction/service/PaymentService.java (hunk 13 above)
Comment (nalijagic1) [reply backend/src/main/java/com/praksa/auction/service/PaymentService.java]: Inline if initialization is redundant

File: backend/src/main/java/com/praksa/auction/model/Person.java (hunk 9 above)
Comment (eminaomanovic) [reply backend/src/main/java/com/praksa/auction/model/Person.java:25]: Please leave a message after dropping this table.

File: backend/src/main/java/com/praksa/auction/service/PaymentService.java (hunk 13 above)
Comment (eminaomanovic) [reply backend/src/main/java/com/praksa/auction/service/PaymentService.java]: How do you mean redundant ? It is Intelij message ?

File: backend/src/main/java/com/praksa/auction/model/Person.java (hunk 9 above)
Comment (nalijagic1) [reply backend/src/main/java/com/praksa/auction/model/Person.java:25]: I dropped the table both locally and on heroku

**Pull request merged**